### PR TITLE
Hunyuan tokenizer

### PR DESCRIPTION
**Pull request by ngxson**
*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

### CODE REVIEW

**Comment by ngxson** [comment]
@coderabbitai pause

File: src/llama-graph.cpp [hunk 1]
```diff
@@ -705,13 +705,6 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
             ggml_reshape_3d(ctx0, probs, 1, n_expert, n_tokens), selected_experts); // [1, n_expert_used, n_tokens]
     cb(weights, "ffn_moe_weights", il);
 
-    if (arch == LLM_ARCH_HUNYUAN_MOE) {
-        weights = ggml_reshape_2d(ctx0, weights, n_expert_used, n_tokens); // [n_expert_used, n_tokens]
-        weights = ggml_div(ctx0, weights, ggml_sum_rows(ctx0, weights)); // [1, n_tokens]
-        weights = ggml_reshape_3d(ctx0, weights, 1, n_expert_used, n_tokens); // [1, n_expert_used, n_tokens]
-        cb(weights, "ffn_moe_weights_scaled", il);
-    }
-
```
Comment (ngxson) [inline src/llama-graph.cpp:714]: haha good catch, I didn't notice that I reinvented the `norm_w` code block

File: gguf-py/gguf/constants.py [hunk 2]
```diff
@@ -148,6 +148,7 @@ class Attention:
         VALUE_LENGTH_MLA             = "{arch}.attention.value_length_mla"
         SHARED_KV_LAYERS             = "{arch}.attention.shared_kv_layers"
         SLIDING_WINDOW_PATTERN       = "{arch}.attention.sliding_window_pattern"
+        QK_NORM                      = "{arch}.attention.qk_norm"
```
Comment (ngxson) [inline gguf-py/gguf/constants.py:151]: rm this

File: convert_hf_to_gguf.py [hunk 3]
```diff
@@ -6420,97 +6425,107 @@ def get_vocab_base(self) -> tuple[list[str], list[int], str]:
...
-        moe_topk = self.hparams["moe_topk"]
+        moe_topk = hparams["moe_topk"]
         assert all(topk == moe_topk[0] for topk in moe_topk)
         self.gguf_writer.add_expert_used_count(moe_topk[0])
 
+        moe_shared_expert = hparams["num_shared_expert"]
+        assert all(n == moe_shared_expert[0] for n in moe_shared_expert)
+        self.gguf_writer.add_expert_shared_count(moe_shared_expert[0])
+
+        self.gguf_writer.add_qk_norm(hparams.get("use_qk_norm", True))
```
Comment (ngxson) [inline convert_hf_to_gguf.py:6481]: rm this

File: convert_hf_to_gguf.py [hunk 4]
```diff
@@ -6420,97 +6425,107 @@ def get_vocab_base(self) -> tuple[list[str], list[int], str]:
...
+
+        # Rope
+        rope_scaling = hparams.get("rope_scaling", {})
+        if rope_scaling.get("type") == "dynamic":
+            self.gguf_writer.add_rope_scaling_type(gguf.RopeScalingType.YARN)
+            self.gguf_writer.add_rope_scaling_factor(rope_scaling["factor"])
```
Comment (ngxson) [inline convert_hf_to_gguf.py:6487]: maybe add an `else: raise Error` here

File: gguf-py/gguf/gguf_writer.py [hunk 5]
```diff
@@ -792,6 +792,9 @@ def add_group_norm_groups(self, value: int) -> None:
     def add_causal_attention(self, value: bool) -> None:
         self.add_bool(Keys.Attention.CAUSAL.format(arch=self.arch), value)
 
+    def add_qk_norm(self, value: bool) -> None:
+        self.add_bool(Keys.Attention.QK_NORM.format(arch=self.arch), value)
+
```
[the remark this answers is not in the document]
Comment (kooshi) [reply gguf-py/gguf/gguf_writer.py]: removed

File: gguf-py/gguf/constants.py (hunk 2 above)
Comment (kooshi) [reply gguf-py/gguf/constants.py]: removed

File: convert_hf_to_gguf.py (hunk 3 above)
Comment (kooshi) [reply convert_hf_to_gguf.py]: removed

File: convert_hf_to_gguf.py (hunk 4 above)
Comment (kooshi) [reply convert_hf_to_gguf.py:6483]: This one I'm not really sure about. It effectively does nothing right now because the factor is 1 in the config. I would need to dig deeper to see if their "dynamic" really is YARN, and I really want to extend it to their claimed 256k. That will require more research after the logits line up.
As for the `raise Error`, it seems other implementations just let it continue if rope is not defined, so I'll leave it as is for now.

**Comment by ubergarm** [comment]
I just tested this PR and re-converted the safetensors to bf16 and seems to be working now in very limited testing.

<details>

<summary>👈 commands and logs</summary>

```bash
## convert
python \
    convert_hf_to_gguf.py \
    --outtype bf16 \
    --split-max-size 50G \
    --outfile /mnt/raid/models/ubergarm/Hunyuan-A13B-Instruct-GGUF/ \
    /mnt/raid/models/tencent/Hunyuan-A13B-Instruct/

## run server
./build/bin/llama-server \
  --model "$model" \
  --alias ubergarm/Hunyuan-A13B-Instruct-bf16 \
  -fa \
  -ctk q8_0 -ctv q8_0 \
  -c 8192 \
  --jinja \
  --temp 0.6 \
  --presence-penalty 0.7 \
  --min-p 0.1 \
  -b 1024 \
  -ts 48,48 \
  -ngl 18 \
  --threads 24 \
  --host 127.0.0.1 \
  --port 8080
```

</details>

Screenshot

mainline-hunyuan-a13b-test
